rfc/20241206-oci-registries/1-oci-primer.md

```diff
@@ -0,0 +1,388 @@
+
+> [!TIP]
+> Try it yourself: authenticate with a token as described above, then use the following command:
+> ```
+> curl -v -H "Authorization: Bearer djE6b3Blb..." https://ghcr.io/v2/opentofu/opentofu/tags/list
+> ```
+> 
+> <details><summary>Results</summary>
+>
+> ```
+> > GET /v2/opentofu/opentofu/tags/list HTTP/2
+> > Host: ghcr.io
+> > User-Agent: curl/8.5.0
+> > Accept: */*
+> > Authorization: Bearer djE6b3B...
+> >
+> 
+> < HTTP/2 200
+> < content-type: application/json
+> < docker-distribution-api-version: registry/2.0
+> < link: </v2/opentofu/opentofu/tags/list?last=1.6.0-beta5-386&n=0>; rel="next"
+> < date: Tue, 04 Feb 2025 09:47:19 GMT
+> < x-github-request-id: D388:2C9949:B77F2:B977F:67A1E226
+> <
+> ```
+> ```json
+> {
+>   "name": "opentofu/opentofu",
+>   "tags": [
+>     "1.6.0-alpha1-arm64",
+>     "1.6.0-alpha1-amd64",
+>     "1.6.0-alpha1-arm",
+>     "1.6.0-alpha1-386",
+>     "1.6.0-alpha1",
+>     "1.6",
+>     "1",
+>     "latest",
+>     "sha256-722da07b0cdf5b6bdf12aff9339f7c274f70552144a0a28c0d2b970c083ffa5c.sig",
+>     "sha256-9b9662dbe859f81779d04ba0f8d2b415b09961d5e5435a25e2ba4566873058f0.sig",
+>     "sha256-d1aa9cfa30744d52a486b1dd58b7a65dd9d1ae9b79fc9b6ee5d0528ef9cfd54c.sig",
+>     "sha256-cd67ab05739a47503450c8d2da4abf89d5418f800d093f43a996de64a0d2fc33.sig",
+>     "sha256-28f8c87d6583f570acf7dd33afa6b17888e7f15ae595d74e16cbd4c82921f0d0.sig",
+>     "..."
+>   ]
+> }
+> ```
+> Note: We omitted the TLS output from this example for readability.
+>
+> </details>
+
+This category also includes a way to create references between distinct manifests. As an example, you can attach a digital signature of a manifest to an existing manifest by publishing the signature under a separate manifest, but referring to the manifest it signs. You can do this by pushing the new manifest with a `subject` field that references the original manifest.
+
+Clients can query the `/v2/<name>/referrers/<reference>` endpoint to receive a list of manifests that refer to the current manifest in their `subject` field.
+
+> [!TIP]
+> The referrers API has been added in the Distribution spec version 1.1. Prominently, Cosign does not appear to use this API to attach signatures. Instead, Cosign creates a tag named after the checksum of the main manifest and suffix it with `.sig`.
+
+## The `_catalog` extension
+
+Although not standardized in the distribution spec, the `/v2/_catalog` endpoint is traditionally used in the [Docker Registry](https://docker-docs.uclv.cu/registry/spec/api/#listing-repositories) as a means to list all images in a registry. However, this endpoint is typically disabled in public registries or only available after authentication.
+
+## ORAS
+
+Everything above refers to the standard container image layout. However, [ORAS](https://oras.land/) describes how artifacts can be stored in a non-standard layout. ORAS today has wide-ranging support.

```

Comment:
Are there any missing details in the [OCI image-spec guidance on artifacts](https://github.com/opencontainers/image-spec/blob/main/manifest.md#guidelines-for-artifact-usage) that you encountered when putting this together?

Reply:
Thanks. The OCI specs were designed by committee so readability suffers in favor of what we can get everyone to agree on. Feel free to mark this one as resolved.